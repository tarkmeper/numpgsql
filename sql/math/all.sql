select aall('{}'::integer[]);
select aall('{1}'::integer[]);
select aall('{0}'::integer[]);
select aall('{1,3,5,1}'::integer[]);
select aall('{1,3,0,1}'::integer[]);
select aall('{{1,3},{0,1}}'::integer[]);

select aall('{true}'::boolean[]);
select aall('{false}'::boolean[]);

