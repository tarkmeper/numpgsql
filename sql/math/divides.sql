SELECT divide('{3,4}'::integer[], '{1,2}'::integer[]);

--Div by 0
SELECT divide('{3,4}'::integer[], '{0,0}'::integer[]);

--Single value
SELECT '{{6,9},{3,4}}'::integer[] / 3;
SELECT divide('{3,4}'::integer[], 0);
