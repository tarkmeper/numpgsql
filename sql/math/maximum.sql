SELECT maximum('{1,4,5,6,7}'::integer[],5);
