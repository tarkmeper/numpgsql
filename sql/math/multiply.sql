SELECT multiply('{1,2}'::integer[], '{3,4}'::integer[]);

--Single value
SELECT '{{1,2},{3,4}}'::integer[] * 3;
