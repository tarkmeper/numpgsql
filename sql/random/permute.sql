SELECT random_seed(1);

SELECT random_permute('{}'::integer[]);
SELECT random_permute('{1,2,3,4,5,6}'::integer[]);
SELECT random_permute('{{1,2,3},{4,5,6}}'::integer[]);

-- Choice without replacement
SELECT random_choice('{}'::integer[], '{3}'::integer[], false);
SELECT random_choice('{1,2,3,4,5}'::integer[], '{3}'::integer[], false);
SELECT random_choice('{1,2,3,4,5}'::integer[], '{2,2}'::integer[], false);
SELECT random_choice('{1,2,3,4,5}'::integer[], '{7}'::integer[], false);
SELECT random_choice('{1,2,3,4,5}'::integer[], '{3,2}'::integer[], false);


-- Choice with replacement
SELECT random_choice('{}'::integer[], '{3}'::integer[], true);
SELECT random_choice('{1,2,3,4,5}'::integer[], '{3}'::integer[], true);
SELECT random_choice('{1,2,3,4,5}'::integer[], '{2,2}'::integer[], true);
SELECT random_choice('{1,2,3,4,5}'::integer[], '{7}'::integer[], true);
SELECT random_choice('{1,2,3,4,5}'::integer[], '{3,2}'::integer[], true);
SELECT random_choice('{1,2,3,4,5}'::integer[], '{3,2,4,2,1,2,3}'::integer[], true);
SELECT random_choice('{1,2,3,4,5}'::integer[], '{{3,2,1,4},{2,1,2,3}}'::integer[], true);
