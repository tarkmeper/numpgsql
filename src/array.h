extern "C"  {
        #include "postgres.h"
        #include "catalog/pg_type.h"
        #include "utils/array.h"
        #include "fmgr.h"
}
#include <vector>
#include <tuple>


inline static size_t arr_cnt(const ArrayType* arr) {
	size_t result = 1;
	for (int i = 0; i < ARR_NDIM(arr); ++i){
		result *= ARR_DIMS(arr)[i];
	}
	return result;
}

//Return values - TODO rename function this was poorly named.
template<class T> inline Datum _ret(const T& val) { T::unimplemented_function; return Datum(); }
template<> inline Datum _ret<double>(const double& val) { PG_RETURN_FLOAT8(val); }
template<> inline Datum _ret<float>(const float& val) { PG_RETURN_FLOAT4(val); }
template<> inline Datum _ret<short>(const short& val) { PG_RETURN_INT16(val); }
template<> inline Datum _ret<int>(const int& val) { PG_RETURN_INT32(val); }
template<> inline Datum _ret<long>(const long& val) { PG_RETURN_INT64(val); }
template<> inline Datum _ret<bool>(const bool& val) { PG_RETURN_BOOL(val); }

//OID types
template<class T> inline Oid oid_type() { return T::unimplemented_feature; }
template<> inline Oid oid_type<float>() { return FLOAT4OID; }
template<> inline Oid oid_type<double>() { return FLOAT8OID; }
template<> inline Oid oid_type<short>() { return INT2OID; }
template<> inline Oid oid_type<int>() { return INT4OID; }
template<> inline Oid oid_type<long>() { return INT8OID; }
template<> inline Oid oid_type<bool>() { return BOOLOID; }

//Alignment codes
template<class T> char oid_align() { return T::unimplemented_feature; }
template<> inline char oid_align<float>() { return 'i'; }
template<> inline char oid_align<double>() { return 'd'; }
template<> inline char oid_align<short>() { return 's'; }
template<> inline char oid_align<int>() { return 'i'; }
template<> inline char oid_align<long>() { return 'd'; }
template<> inline char oid_align<bool>() { return 'c'; }

//Get param values
template<class T> inline T get_param(PG_FUNCTION_ARGS, int p)  { T::unimplemented_function; }
template<> inline bool get_param<bool>(PG_FUNCTION_ARGS, int p) { return PG_GETARG_BOOL(p); }
template<> inline float get_param<float>(PG_FUNCTION_ARGS, int p) { return PG_GETARG_FLOAT4(p); }
template<> inline double get_param<double>(PG_FUNCTION_ARGS, int p) { return PG_GETARG_FLOAT8(p); }
template<> inline short get_param<short>(PG_FUNCTION_ARGS, int p) { return PG_GETARG_INT16(p); }
template<> inline int get_param<int>(PG_FUNCTION_ARGS, int p) { return PG_GETARG_INT32(p); }
template<> inline long get_param<long>(PG_FUNCTION_ARGS, int p) { return PG_GETARG_INT64(p); }
