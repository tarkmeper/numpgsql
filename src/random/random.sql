-- This program is open source, licensed under the MIT License.
-- For license terms, see the LICENSE file.
--
-- Copyright (C) 2019: Mark Teper

-- Other functions

CREATE FUNCTION random_seed(integer) RETURNS void LANGUAGE C STRICT COST 100 AS '$libdir/numpgsql', 'set_seed';
CREATE FUNCTION random_state() RETURNS text LANGUAGE C STRICT COST 100 AS '$libdir/numpgsql', 'get_rgen_state';
CREATE FUNCTION random_state(text) RETURNS void LANGUAGE C STRICT COST 100 AS '$libdir/numpgsql', 'set_rgen_state';

CREATE FUNCTION random_rand(VARIADIC integer[]) RETURNS double precision[] LANGUAGE C STRICT COST 100 AS '$libdir/numpgsql', 'gen_rand';
CREATE FUNCTION random_randn(VARIADIC integer[]) RETURNS double precision[] LANGUAGE C STRICT COST 100 AS '$libdir/numpgsql', 'gen_randn';
CREATE FUNCTION random_randint(low bigint, high bigint, integer[]) RETURNS bigint[] LANGUAGE C STRICT COST 100 AS '$libdir/numpgsql', 'gen_randint';

CREATE FUNCTION random_permute(anyarray) RETURNS anyarray LANGUAGE C STRICT COST 100 AS '$libdir/numpgsql', 'permute';
CREATE FUNCTION random_choice(anyarray, shape int[], replace boolean) RETURNS anyarray LANGUAGE C STRICT COST 100 AS '$libdir/numpgsql', 'choice';


CREATE FUNCTION random_beta(alpha double precision, beta double precision, shape int[]) RETURNS double precision[] LANGUAGE C STRICT COST 100 AS '$libdir/numpgsql', 'gen_beta';
CREATE FUNCTION random_binomial(t int, p double precision, shape int[]) RETURNS integer[] LANGUAGE C STRICT COST 100 AS '$libdir/numpgsql', 'gen_binomial';

