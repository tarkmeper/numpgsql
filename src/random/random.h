#ifndef RANDOM_H
#define RANDOM_H

#include <random>

extern std::mt19937 rgen;

struct ArrDims {
   int ndims;
   int* dims;
   int lbs[MAXDIM];
   size_t n;
};

extern ArrDims extract_shape(ArrayType* arr);

#endif
