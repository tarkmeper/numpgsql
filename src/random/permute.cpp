extern "C"  {
        #include "postgres.h"
        #include "catalog/pg_type.h"
        #include "fmgr.h"
        #include "math.h"
        #include <utils/lsyscache.h>
}
#include "random.h"
#include "../array.h"

#include <algorithm>


extern "C" { PG_FUNCTION_INFO_V1(permute); Datum permute(PG_FUNCTION_ARGS); }
Datum permute(PG_FUNCTION_ARGS) {
  ArrayType* arr = PG_GETARG_ARRAYTYPE_P(0);
  Oid        type  = ARR_ELEMTYPE(arr);
  int16      elmlen;
  bool       elmbyval;
  char       elmalign;
  Datum*     elemsp;
  bool*      nullsp;
  int        nelemsp;

  if (ARR_NDIM(arr) == 0 ) {
     PG_RETURN_POINTER(arr);
  } else if (ARR_NDIM(arr) != 1 ) {
     elog(ERROR, "PERMUTE: Currently only supports 1d arrays");
     PG_RETURN_NULL();
  }

  const size_t n = ARR_DIMS(arr)[0];
  get_typlenbyvalalign(type, &elmlen, &elmbyval, &elmalign);
  deconstruct_array(arr, type, elmlen, elmbyval, elmalign, &elemsp, &nullsp, &nelemsp);

  Datum* rev = (Datum*)palloc(n * sizeof(Datum));
  std::copy(elemsp, elemsp + n, rev);
  std::shuffle(rev, rev+n, rgen);

  ArrayType* res = construct_array(rev, n, type, elmlen, elmbyval, elmalign);
  PG_RETURN_POINTER(res);
}

// Choice.
extern "C" { PG_FUNCTION_INFO_V1(choice); Datum choice(PG_FUNCTION_ARGS); }
Datum choice(PG_FUNCTION_ARGS) {
  ArrayType* arr = PG_GETARG_ARRAYTYPE_P(0);
  ArrayType* idx = PG_GETARG_ARRAYTYPE_P(1);
  bool replace = get_param<bool>(fcinfo, 2);
  Oid        type  = ARR_ELEMTYPE(arr);
  int16      elmlen;
  bool       elmbyval;
  char       elmalign;
  Datum*     elemsp;
  bool*      nullsp;
  int        nelemsp;

  //cehck for errors etc.
  if (ARR_NDIM(arr) > 1) {
     elog(ERROR, "CHOICE: Currently only supports 1d arrays");
     PG_RETURN_NULL();
  } else if (ARR_NDIM(arr) == 0 ) {
     elog(ERROR, "CHOICE: Input array must ahve at least 1 element");
     PG_RETURN_NULL();
  } else if (ARR_NDIM(idx) != 1 ) {
     elog(ERROR, "CHOICE: Size array must be 1d array");
     PG_RETURN_NULL();
  } else if (ARR_DIMS(idx)[0] >= MAXDIM) {
     elog(ERROR, "CHOICE: Multi-dimension array must have less than %d dimensions has %d", MAXDIM, ARR_DIMS(idx)[0]);
     PG_RETURN_NULL();
  }

  const size_t input_n = ARR_DIMS(arr)[0];
  get_typlenbyvalalign(type, &elmlen, &elmbyval, &elmalign);
  deconstruct_array(arr, type, elmlen, elmbyval, elmalign, &elemsp, &nullsp, &nelemsp);

  ArrDims rd = extract_shape(idx);

  Datum* r = (Datum*)palloc(rd.n * sizeof(Datum));

  if (replace) {
     std::uniform_int_distribution<> dist(0, input_n-1);
     for (size_t i = 0; i < rd.n; ++i ) {
        r[i] = elemsp[dist(rgen)];
     }
  } else {
      if ( rd.n > input_n) { 
        elog(ERROR, "Without replacement target vector cannot be better then input");
        PG_RETURN_NULL();
      }
      std::vector<Datum> tmp(elemsp, elemsp + nelemsp);
      std::shuffle(tmp.begin(), tmp.end(), rgen);
      std::copy(tmp.begin(), tmp.begin() + rd.n, r);
  }

  ArrayType* res = construct_md_array(r,
                                      NULL,
                                      rd.ndims,
                                      rd.dims,
                                      rd.lbs,
                                      type,
                                      elmlen,
                                      elmbyval,
                                      elmalign);


  PG_RETURN_POINTER(res);
}

