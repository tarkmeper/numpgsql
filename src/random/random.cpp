extern "C"  {
        #include "postgres.h"
        #include "catalog/pg_type.h"
        #include "fmgr.h"
        #include "math.h"
}
#include <string>
#include <iostream>
#include <sstream>
#include "random.h"
#include "../array.h"

std::mt19937 rgen;

ArrDims extract_shape(ArrayType* arr) {
  static constexpr int LBS_VALS[] = { 1, 1, 1, 1, 1, 1 };
  ArrDims res;
  res.dims = (int*)(ARR_DATA_PTR(arr));
  res.ndims = ARR_DIMS(arr)[0];
  res.n = res.dims[0];
  memcpy(res.lbs, LBS_VALS, sizeof(int) * MAXDIM);
  for ( int i = 1; i < res.ndims; ++i) { res.n *= res.dims[i]; }
  return res;
}

extern "C" { PG_FUNCTION_INFO_V1(set_seed); Datum set_seed(PG_FUNCTION_ARGS); }
 Datum set_seed(PG_FUNCTION_ARGS) {
  const int s = PG_GETARG_INT32(0);
  rgen.seed(s);
  PG_RETURN_NULL();
}


extern "C" { PG_FUNCTION_INFO_V1(get_rgen_state); Datum get_rgen_state(PG_FUNCTION_ARGS); }
Datum get_rgen_state(PG_FUNCTION_ARGS) {
  std::stringstream ss;
  ss << rgen;

  const std::string val = ss.str();

  //Now move over
  text *new_t = (text *) palloc(val.size() + VARHDRSZ);
  SET_VARSIZE(new_t, val.size() + VARHDRSZ);
  memcpy(VARDATA(new_t), val.c_str(), val.size());

  PG_RETURN_TEXT_P(new_t);
}


extern "C" { PG_FUNCTION_INFO_V1(set_rgen_state); Datum set_rgen_state(PG_FUNCTION_ARGS); }
Datum set_rgen_state(PG_FUNCTION_ARGS) {
  text  *arg1 = PG_GETARG_TEXT_PP(0);  std::string val(VARDATA(arg1));
  std::istringstream ss(val);
  ss >> rgen;

  PG_RETURN_NULL();
}


template<class D> static Datum _gen_rand(PG_FUNCTION_ARGS, D& dist, int dim_arr_idx = 0) {
  typedef typename D::result_type T;
  ArrayType* arr = PG_GETARG_ARRAYTYPE_P_COPY(dim_arr_idx);

  if (ARR_NDIM(arr) == 0 ) {
     PG_RETURN_POINTER(construct_empty_array(oid_type<double>()));
  } else if (ARR_NDIM(arr) != 1 ) {
     elog(ERROR, "GEN_RAND: Size specification can only be 1-dimension array");
     PG_RETURN_NULL();
  } else if (ARR_DIMS(arr)[0] >= 4 ) {
     elog(ERROR, "GEN_RAND: Maximum of 4 dimensions in array generation.");
     PG_RETURN_NULL();
  }

  //compute number of dimensions
  ArrDims rd = extract_shape(arr);
  Datum* r = (Datum*)palloc(rd.n * sizeof(Datum));
  for (size_t i = 0; i < rd.n; ++i ) {
     r[i] = _ret<T>(dist(rgen));
  }

  ArrayType* res = construct_md_array(r,
                                      NULL,
                                      rd.ndims,
                                      rd.dims,
                                      rd.lbs,
                                      oid_type<T>(),
                                      sizeof(T),
                                      true,
                                      oid_align<T>());
  PG_RETURN_POINTER(res);
}

extern "C" { PG_FUNCTION_INFO_V1(gen_rand); Datum gen_rand(PG_FUNCTION_ARGS); }
Datum gen_rand(PG_FUNCTION_ARGS) {
  std::uniform_real_distribution<double> dist{0.0, 1.0};
  return _gen_rand(fcinfo, dist);
}


extern "C" { PG_FUNCTION_INFO_V1(gen_randn); Datum gen_randn(PG_FUNCTION_ARGS); }
Datum gen_randn(PG_FUNCTION_ARGS) {
  std::normal_distribution<double> dist{0.0, 1.0};
  return _gen_rand(fcinfo, dist);
}


extern "C" { PG_FUNCTION_INFO_V1(gen_randint); Datum gen_randint(PG_FUNCTION_ARGS); }
Datum gen_randint(PG_FUNCTION_ARGS) {
  long low = get_param<long>(fcinfo, 0);
  long high = get_param<long>(fcinfo, 1);
  if (high <= low ){
     elog(ERROR, "GEN_RANDINT: Cannot have high bound lower then low bound");
     PG_RETURN_NULL();
  } else {
     std::uniform_int_distribution<> dist(low, high);
     return _gen_rand(fcinfo, dist, 2);
  }
}
