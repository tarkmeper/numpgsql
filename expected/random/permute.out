SELECT random_seed(1);
 random_seed 
-------------
 
(1 row)

SELECT random_permute('{}'::integer[]);
 random_permute 
----------------
 {}
(1 row)

SELECT random_permute('{1,2,3,4,5,6}'::integer[]);
 random_permute 
----------------
 {6,1,4,3,5,2}
(1 row)

SELECT random_permute('{{1,2,3},{4,5,6}}'::integer[]);
ERROR:  PERMUTE: Currently only supports 1d arrays
-- Choice without replacement
SELECT random_choice('{}'::integer[], '{3}'::integer[], false);
ERROR:  CHOICE: Input array must ahve at least 1 element
SELECT random_choice('{1,2,3,4,5}'::integer[], '{3}'::integer[], false);
 random_choice 
---------------
 {5,1,2}
(1 row)

SELECT random_choice('{1,2,3,4,5}'::integer[], '{2,2}'::integer[], false);
 random_choice 
---------------
 {{5,4},{2,1}}
(1 row)

SELECT random_choice('{1,2,3,4,5}'::integer[], '{7}'::integer[], false);
ERROR:  Without replacement target vector cannot be better then input
SELECT random_choice('{1,2,3,4,5}'::integer[], '{3,2}'::integer[], false);
ERROR:  Without replacement target vector cannot be better then input
-- Choice with replacement
SELECT random_choice('{}'::integer[], '{3}'::integer[], true);
ERROR:  CHOICE: Input array must ahve at least 1 element
SELECT random_choice('{1,2,3,4,5}'::integer[], '{3}'::integer[], true);
 random_choice 
---------------
 {2,2,4}
(1 row)

SELECT random_choice('{1,2,3,4,5}'::integer[], '{2,2}'::integer[], true);
 random_choice 
---------------
 {{2,5},{3,5}}
(1 row)

SELECT random_choice('{1,2,3,4,5}'::integer[], '{7}'::integer[], true);
  random_choice  
-----------------
 {3,2,4,3,2,3,5}
(1 row)

SELECT random_choice('{1,2,3,4,5}'::integer[], '{3,2}'::integer[], true);
    random_choice    
---------------------
 {{2,1},{3,4},{5,3}}
(1 row)

SELECT random_choice('{1,2,3,4,5}'::integer[], '{3,2,4,2,1,2,3}'::integer[], true);
ERROR:  CHOICE: Multi-dimension array must have less than 6 dimensions has 7
SELECT random_choice('{1,2,3,4,5}'::integer[], '{{3,2,1,4},{2,1,2,3}}'::integer[], true);
ERROR:  CHOICE: Size array must be 1d array
